test/lit/wasm-split/asyncify.wast

```diff
@@ -0,0 +1,174 @@
+;; RUN: wasm-split %s --export-prefix='%' -g -o1 %t.1.wasm -o2 %t.2.wasm --keep-funcs=foo --asyncify
+;; RUN: wasm-dis %t.1.wasm | filecheck %s --check-prefix PRIMARY
+;; RUN: wasm-dis %t.2.wasm | filecheck %s --check-prefix SECONDARY
+
+;; Check that the --asyncify option instruments the primary module but not the
+;; secondary module.
+
+(module
+ (func $foo (param i32) (result i32)
+  (call $bar (i32.const 0))
+ )
+ (func $bar (param i32) (result i32)
+  (call $foo (i32.const 1))
+ )
+)
+
+;; PRIMARY:      (module
+;; PRIMARY-NEXT:  (type $i32_=>_i32 (func (param i32) (result i32)))
+;; PRIMARY-NEXT:  (type $i32_=>_none (func (param i32)))
+;; PRIMARY-NEXT:  (type $none_=>_none (func))
+;; PRIMARY-NEXT:  (type $none_=>_i32 (func (result i32)))
+;; PRIMARY-NEXT:  (import "placeholder" "0" (func $placeholder_0 (param i32) (result i32)))
+;; PRIMARY-NEXT:  (global $global$0 (mut i32) (i32.const 0))
+;; PRIMARY-NEXT:  (global $global$1 (mut i32) (i32.const 0))
+;; PRIMARY-NEXT:  (memory $0 1 1)
+;; PRIMARY-NEXT:  (table $0 1 funcref)
+;; PRIMARY-NEXT:  (elem (i32.const 0) $placeholder_0)
+;; PRIMARY-NEXT:  (export "%foo" (func $foo))
+;; PRIMARY-NEXT:  (export "%table" (table $0))
+;; PRIMARY-NEXT:  (export "asyncify_start_unwind" (func $asyncify_start_unwind))
+;; PRIMARY-NEXT:  (export "asyncify_stop_unwind" (func $asyncify_stop_unwind))
+;; PRIMARY-NEXT:  (export "asyncify_start_rewind" (func $asyncify_start_rewind))
+;; PRIMARY-NEXT:  (export "asyncify_stop_rewind" (func $asyncify_stop_rewind))
+;; PRIMARY-NEXT:  (export "asyncify_get_state" (func $asyncify_get_state))
+;; PRIMARY-NEXT:  (func $foo (param $0 i32) (result i32)
```

Comment:
This is a shocking amount of code to instrument a single indirect call. Is there room for improvement here? It looks like there is a stack pointer being manipulated, probably into the stack of spilled locals, but this function doesn't need to spill any locals (AFAIK), so I'm surprised that code is generated.

Reply:
Yeah, quite a lot...

This is with both wasm-split and asyncify, though - do you have the code after just wasm-split? I think asyncify is seeing more than just a single call, as it seems to be spilling locals - does wasm-split add any?

Reply:
This is the code after just wasm-split:
```
 (func $foo (param $0 i32) (result i32)
  (call_indirect (type $i32_=>_i32)
   (i32.const 0)
   (i32.const 0)
  )
 )
 ```

Reply:
Ah, ok, thanks...

Reading the output more carefully, it could indeed be better. It is handling the case of a return value from a call carefully by adding another local, which forces it to add spilling code. That doesn't matter here I think since the code is so trivial, but it doesn't take advantage of that. Second, it writes and reads the "pause position" index in the function, but there is just a single possible position here, so that code is redundant as well - again, we don't have a specific optimization to benefit from the case of a single possible pause site, so the code is more general than it needs to be...